go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -184,18 +341,97 @@ func (npw *nodePortWatcher) SyncServices(services []interface{}) {
+			if countHostNeworkedEndpoints(ep, npw.nodeName) == 0 {
+				klog.V(5).Infof("Service %s has ETP=local and no local Endpoints, Updating flow Cache", ep.Name)
+				npw.updateServiceFlowCache(npw.services[name], false, true)
+				npw.ofm.requestFlowSync()
+				delSharedGatewayIptRules(npw.services[name], true, npw.nodeIPs)
+			}
+		}
+	}
+	// We add this only when a Service event is received, delete only when endpoint delete event is received
+	delete(npw.services, name)
+}
+
+func (npw *nodePortWatcher) UpdateEndpoints(old, new *kapi.Endpoints) {
+	name := ktypes.NamespacedName{Namespace: new.Namespace, Name: new.Name}
+	if _, exists := npw.services[name]; exists {
+		if util.ServiceExternalTrafficPolicyLocal(npw.services[name]) {
+			needFlowSync := false
+			// We had special ETP flows but now we don't
+			if countHostNeworkedEndpoints(old, npw.nodeName) > 0 && countHostNeworkedEndpoints(new, npw.nodeName) == 0 {
+				klog.V(5).Infof("Service %s has ETP=local, had local Endpoints, but now does not Updating flow Cache", new.Name)
+				// Delete ETP flows
+				npw.updateServiceFlowCache(npw.services[name], false, true)
+				needFlowSync = true
+				// Delete Special Iptables Rules
+				delSharedGatewayIptRules(npw.services[name], true, npw.nodeIPs)
+				// If There are still non-local eps Re-sync with normal nodeport service flows
+				if len(new.Subsets) > 0 {
+					klog.V(5).Infof("Service %s has ETP=local, had local Endpoints, and still has normal ones Updating flow Cache", new.Name)
+					// Update flow cache with standard flows
+					npw.updateServiceFlowCache(npw.services[name], true, false)
+					addSharedGatewayIptRules(npw.services[name], false, npw.nodeIPs)
+				}
+			}
+			// Add Special ETP flows if there are now host endpoints
+			if countHostNeworkedEndpoints(old, npw.nodeName) == 0 && countHostNeworkedEndpoints(new, npw.nodeName) > 0 {
+				// Update flow cache with special ones
+				klog.V(5).Infof("Service %s has ETP=local and didn't have local Endpoints but now does Updating flow Cache", new.Name)
+				npw.updateServiceFlowCache(npw.services[name], true, true)
+				needFlowSync = true
+				// delete standard IpTRules to add special ones
+				delSharedGatewayIptRules(npw.services[name], false, npw.nodeIPs)
+				addSharedGatewayIptRules(npw.services[name], true, npw.nodeIPs)
+			}
+
+			if needFlowSync {
+				npw.ofm.requestFlowSync()
+			}
+		}
+
+	}
 }
 
-// since we share the host's k8s node IP, add OpenFlow flows
+// since we share the host's k8s node IP, add OpenFlow flows to br-ex
 // -- to steer the NodePort traffic arriving on the host to the OVN logical topology and
 // -- to also connection track the outbound north-south traffic through l3 gateway so that
 //    the return traffic can be steered back to OVN logical topology
 // -- to handle host -> service access, via masquerading from the host to OVN GR
+// -- to handle host -> service(ExternalTrafficPolicy: Local) -> host access without SNAT
```

Comment:
this isn't really true, is it? I thought we are using iptables rules so that host -> external traffic policy local traffic is directly DNAT'ed locally and does not enter the bridge?

Reply:
Yep, The flows here handle ingress traffic, will fix the comment